great_expectations/expectations/metrics/util.py

```diff
@@ -1058,3 +1060,165 @@ def get_sqlalchemy_source_table_and_schema(
         )
     else:
         return engine.batch_manager.active_batch_data.selectable
+
+
+def get_unexpected_indices_for_multiple_pandas_named_indices(
+    domain_records_df: pd.DataFrame,
+    unexpected_index_column_names: List[str],
+    expectation_domain_column_name: Union[str, None] = None,
+) -> List[Dict[str, Any]]:
+    """
+    Builds unexpected_index list for Pandas Dataframe in situation where the named
+    columns is also a named index. This method handles the case when there are multiple named indices.
+    Args:
+        domain_records_df: reference to Pandas dataframe
+        expectation_domain_column_name: column that Expectation is being run for
+        unexpected_index_column_names: column_names for indices, either named index or unexpected_index_columns
+
+    Returns:
+        List of Dicts that contain ID/PK values
+    """
+    if expectation_domain_column_name is None:
+        raise ge_exceptions.MetricResolutionError(
+            message=f"Error: The domain column is currently set to None. Please check your configuration.",
+            failed_metrics=["unexpected_index_list"],
+        )
+
+    domain_records_df_index_names: List[str] = domain_records_df.index.names
+    unexpected_indices: List[Union[int, str]] = list(domain_records_df.index)
+
+    for column_name in unexpected_index_column_names:
+        if column_name not in domain_records_df_index_names:
+            raise ge_exceptions.MetricResolutionError(
+                message=f"Error: The column {column_name} does not exist in the named indices. Please check your configuration",
+                failed_metrics=["unexpected_index_list"],
+            )
+
+    unexpected_index_list: List[Dict[str, Any]] = list()
+
+    for index in unexpected_indices:
+        primary_key_dict: Dict[str, Any] = dict()
+        # domain column first
+        primary_key_dict[expectation_domain_column_name] = domain_records_df.at[
+            index, expectation_domain_column_name
+        ]
+        for column_name in unexpected_index_column_names:
+            tuple_index = domain_records_df_index_names.index(column_name, 0)
```

Comment:
You may want to define `tuple_index` outside this `for` loop and incorporate it into the `for column_name in unexpected_index_column_names` loop above since the `index` call will loop through the list `domain_records_df_index_names` looking for a match time it is invoked. However, depending on the size of the `domain_records_df_index_names` and the speed of hashing I'm not sure what will be faster so maybe it doesn't matter.

Reply:
I had to think a little bit more about this, but I think it's a great suggestion, even with the uncertainty around the speed of hashing. I'd adjusted the loop above to build a dictionary of the column names and tuple indices. That away this line becomes a much simple dictionary value retrieval operation 🙇🏼 thanks for this